reconciler/reconcile_common_test.go

```diff
@@ -0,0 +1,128 @@
+		},
+
+		Status: v1.Status{
+			ObservedGeneration: 0,
+			Conditions:         v1.Conditions{*foo, *bar},
+		},
+	}
+
+	new := &v1.KResource{
+		ObjectMeta: metav1.ObjectMeta{
+			Generation: 1,
+		},
+
+		Status: v1.Status{
+			ObservedGeneration: 0,
+			Conditions:         v1.Conditions{*foo, *bar},
+		},
+	}
+	return old, new
+}
+
+func TestPostProcessReconcileBumpsGeneration(t *testing.T) {
+	old, new := makeResources()
+
+	oldShape := v1.KRShaped(old)
+	newShape := v1.KRShaped(new)
+	PostProcessReconcile(context.Background(), oldShape, newShape, nil)
+
+	if new.Status.ObservedGeneration != new.Generation {
+		t.Errorf("Expected observed generation bump got=%d want=%d", new.Status.ObservedGeneration, new.Generation)
+	}
+
+	if newShape.GetStatus().ObservedGeneration != newShape.GetObjectMeta().Generation {
+		t.Errorf("Expected observed generation bump got=%d want=%d", new.Status.ObservedGeneration, new.Generation)
+	}
+}
+
+func TestPostProcessReconcileBumpsWithEvent(t *testing.T) {
+	old, new := makeResources()
+	reconcilEvent := NewEvent(corev1.EventTypeWarning, exampleStatusFailed, "")
+
+	oldShape := v1.KRShaped(old)
+	newShape := v1.KRShaped(new)
+	PostProcessReconcile(context.Background(), oldShape, newShape, reconcilEvent)
+
+	// Expect generation bumped
+	if new.Status.ObservedGeneration != new.Generation {
+		t.Errorf("Expected observed generation bump got=%d want=%d", new.Status.ObservedGeneration, new.Generation)
+	}
+
+	if newShape.GetStatus().ObservedGeneration != newShape.GetObjectMeta().Generation {
+		t.Errorf("Expected observed generation bump got=%d want=%d", new.Status.ObservedGeneration, new.Generation)
+	}
+
+	// The old/new conditions were not changed, expect sets unknown
+	if rc := new.Status.GetCondition(apis.ConditionReady); rc.Status != "Unknown" {
+		t.Errorf("Expected unknown ready status got=%s", rc.Status)
+	}
+}
+
+func TestPostProcessWithEventCondChange(t *testing.T) {
+	old, new := makeResources()
+	originalReadyStatus := old.Status.GetCondition(apis.ConditionReady).Status
+	old.Status.Conditions = make([]apis.Condition, 0)
+	reconcilEvent := NewEvent(corev1.EventTypeWarning, exampleStatusFailed, "")
+
+	oldShape := v1.KRShaped(old)
+	newShape := v1.KRShaped(new)
+	PostProcessReconcile(context.Background(), oldShape, newShape, reconcilEvent)
+
+	// Expect generation bumped
+	if new.Status.ObservedGeneration != new.Generation {
+		t.Errorf("Expected observed generation bump got=%d want=%d", new.Status.ObservedGeneration, new.Generation)
+	}
+
+	if newShape.GetStatus().ObservedGeneration != newShape.GetObjectMeta().Generation {
+		t.Errorf("Expected observed generation bump got=%d want=%d", new.Status.ObservedGeneration, new.Generation)
+	}
+
+	// The old/new conditions were changed, expect that ready remains unchanged
+	if rc := new.Status.GetCondition(apis.ConditionReady); rc.Status != originalReadyStatus {
+		t.Errorf("Expected unchanged ready status got=%s want=%s", rc.Status, originalReadyStatus)
+	}
+}
```

Comment:
I think a table test would be a good fit here:
```go
testCases := []struct {
    name           string
    reconcileEvent *Event
    newCondition   *apis.Condition
}{
    {
        name:           "sets Unknown condition if returned event and no condition change",
        reconcileEvent: NewEvent(corev1.EventTypeWarning, exampleStatusFailed, ""),
    },
    {
        name:           "leaves condition unchanged if returned event and condition changed",
        reconcileEvent: NewEvent(corev1.EventTypeWarning, exampleStatusFailed, ""),
        newCondition:   &apis.Condition{Type: apis.ConditionReady, Status: corev1.ConditionFalse},
    },
}
```

Reply:
I've changed this a bunch - should be much simpler now.